src/engraving/libmscore/layout.cpp

```diff
@@ -2156,6 +2156,9 @@ System* Score::getNextSystem(LayoutContext& lc)
     if (!isVBox) {
         int nstaves = Score::nstaves();
         system->adjustStavesNumber(nstaves);
+        for (int i = 0; i < nstaves; ++i) {
+            system->staff(i)->setShow(score()->staff(i)->show());
+        }
```

Comment:
without this, it "remembers" which staves were hidden for this system on the previous layout, but that changes whether it takes into account the instrument name when calculating the width, meaning you can get a different result depending on the state of the cache.

Reply:
This change, backported to the 3.x branch, causes MuseScore(3.exe) to crash (in Debug builds, QtCreator, MinGW)), on  opening any score due to a failed assertion in Qt:
Fatal: ASSERT failure in QList<T>::operator[]: "index out of range", file C:/Qt/5.15.2/mingw81_64/include/QtCore/qlist.h, line 575

Not sure how relevant this is to the master branch though.

Reply:
I worked this fix out in MU3 and if it'd crashing there (which it wasn't for me) it's wrong in MU4 too - I'll have a look when I can.

Reply:
It certainly crashes for me.

Reply:
Oops, I added these 3 lines a line too late (before the `system->adjustStavesNumber(nstaves);` rather than after), downside of an all-manual cherry-pick. Have a brown bag for me?
Sorry for the noise...

Reply:
Yeah I was pretty sure it was OK!  Took me hours to work out what that problem was too BTW, surprised it was such a simple change in the end, and a bit surprised it doesn't cause more weird layout issues.